src/nifs/protogalaxy/pow_i.rs

```diff
@@ -0,0 +1,123 @@
+use std::{iter, num::NonZeroUsize};
+
+use bitter::{BitReader, LittleEndianReader};
+use ff::PrimeField;
+use tracing::*;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Input `t` is not enought for represent `i`")]
+    TooLittleT {
+        t: NonZeroUsize,
+        representation_lenght: NonZeroUsize,
+    },
+}
+
+fn pow_i<F: PrimeField>(
+    i: usize,
+    t: NonZeroUsize,
+    challenges_powers: impl Iterator<Item = F>,
+) -> F {
+    let bytes = i.to_le_bytes();
+    let mut reader = LittleEndianReader::new(&bytes);
+
+    iter::repeat_with(|| reader.read_bit().unwrap_or(false))
+        .zip(challenges_powers)
+        .map(|(b_j, beta_in_2j)| match b_j {
+            true => beta_in_2j,
+            false => F::ONE,
+        })
+        .take(t.get())
+        .reduce(|acc, coeff| acc * coeff)
+        .unwrap()
+}
+
+/// Constructs an iterator yielding evaluated `pow_i` for indices from 0 to `n`.
+///
+/// # Params
+/// - `t` - lenght of binary representation of `i`. If `t` is too small to represent `i`, then
+///       [`PowError::TooLittleT`] will return.
+/// - `n` - limit for `pow_i` relations
+/// - `challenge` - challenge for `pow_i` relation
+///
+/// # Mathematical Representation
+///
+/// For an integer `i` ranging from 0 to `n`, its binary representation `[i]_2 = (b_0, b_1, ..., b_{t-1})`,
+/// and an input expression `challenge`, the polynomial term `pow_i` is defined as:
+///
+/// `beta` is `challenge`
+/// ```math
+/// pow_i(\boldsymbol{\beta}) =
+///     pow_i(\beta, \cdots, \beta^{2^{t-1}}) =
+///     \prod_{j=0}^{t-1}(1 - b_j + b_j \cdot \beta^{2^j})
+/// ```
+/// where `b_j` is the j-th bit of the binary representation of the integer `i`.
+pub fn iter_eval_of_pow_i<F: PrimeField>(
+    t: NonZeroUsize,
+    n: NonZeroUsize,
+    challenge: F,
+) -> Result<impl Iterator<Item = F>, Error> {
+    let n = n.get();
+
+    if t.get() < n.leading_ones() as usize {
```

Comment:
`t=ceil(log(n))`.   Suppose `[n]_2=(1000), n.leading_ones()=1` (if my understanding of this fn is correct), then t=2 is not enough.

Reply:
Yes, the function is used incorrectly. I replaced it with `mem::size_of<usize>() * 8 - n.leading_zeros()`. That is, it counts how many bits are needed to represent a number. I also added a test to check this